src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
-                        if (IsStar || IsPlus)
+                        // Helper function for the case where the right side consumes the character
+                        static bool RightTransition(SymbolicRegexNode<S> node, S elem, uint context, List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions,
+                            List<SymbolicRegexNode<S>> continuation, Stack<DerivativeEffect>? effects, bool eager)
                         {
-                            transition = step.Concat(_builder.CreateLoop(_left, IsLazy));
+                            Debug.Assert(node._left is not null && node._right is not null);
+                            // Remember current number of effects so that we know how many to pop
+                            int j = effects?.Count ?? 0;
+                            if (effects is not null)
+                            {
+                                // Push all effects onto the effects stack
+                                node._left.ApplyEffects(effect => effects.Push(effect), context);
+                            }
+                            bool done = node._right.AddTransitions(elem, context, transitions, continuation, effects, eager);
+                            if (!done && effects is not null)
+                            {
+                                // Pop all effects that were added here
+                                while (effects.Count > j)
+                                    effects.Pop();
+                            }
+                            return done;
                         }
-                        else
+
+                        // Helper function for the case where the left side consumes the character
+                        static bool LeftTransition(SymbolicRegexNode<S> node, S elem, uint context, List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions,
+                            List<SymbolicRegexNode<S>> continuation, Stack<DerivativeEffect>? effects, bool eager)
                         {
-                            int newupper = _upper == int.MaxValue ? int.MaxValue : _upper - 1;
-                            int newlower = _lower == 0 ? 0 : _lower - 1;
-                            SymbolicRegexNode<S> rest = _builder.CreateLoop(_left, IsLazy, newlower, newupper);
-                            transition = step.Concat(rest);
+                            Debug.Assert(node._left is not null && node._right is not null);
+                            continuation.Add(node._right);
+                            // Make the left side non-eager, i.e. consider all paths, if the right side is not nullable here.
+                            // The intuition is that if the right side is nullable, then backtracking would not accept the empty
+                            // string here even when it hits a nullable path for the left side, so all paths through the left side
+                            // need to be considered.
```

Comment:
It was great earlier when you had concrete examples to explain the concept, e.g. where you showed a pattern and what it's various derivatives would be.  For a bunch of these comments that would also be very useful.

Reply:
I added a few examples around that function.